packages/sdk/src/oracle/services/statemachines/statemachine.ts

```diff
@@ -0,0 +1,106 @@
+ * 5. Context.done - This is a special property on context.done which represents that the state machine is done transitioning this context.
+ * 6. Context.state = "done" - This is a reserved state on context.state, if set to "done" its the same thing as done = true.
+ * 7. Context.state = "error" - This is a reserved state on context.state, if set to "error" it means the context is done, but also there is an context.error object.
+ * 8. Interval - This is a property on the context which specifies the maximum rate in MS this context will transition, based on the current timestamp.
+ */
+export class StateMachine {
+  private pending: Context<unknown, unknown & Memory>[] = [];
+  // this needs to be updated when adding new context type
+  public types: {
+    [ContextType.setUser]: ContextManager<setUser.Params, setUser.Memory>;
+    [ContextType.setActiveRequest]: ContextManager<setActiveRequest.Params, setActiveRequest.Memory>;
+  };
+  constructor(private store: Store) {
+    // need to initizlie state types here manually for each new context type
+    this.types = {
+      [ContextType.setUser]: new ContextManager<setUser.Params, setUser.Memory>(
+        ContextType.setUser,
+        setUser.Handlers(store),
+        setUser.initMemory,
+        this.handleCreate
+      ),
+      [ContextType.setActiveRequest]: new ContextManager<setActiveRequest.Params, setActiveRequest.Memory>(
+        ContextType.setActiveRequest,
+        setActiveRequest.Handlers(store),
+        setActiveRequest.initMemory,
+        this.handleCreate
+      ),
+    };
+  }
+  private saveContext(context: Context<unknown, unknown & Memory>) {
+    this.store.write((w) => w.command(context));
+  }
+  private handleCreate = (context: Context<unknown, unknown & Memory>) => {
+    this.pending.push(context);
+    this.saveContext(context);
+  };
+
+  private pop(): Context<unknown, unknown & Memory> | undefined {
+    return this.pending.pop();
+  }
+  private size(): number {
+    return this.pending.length;
+  }
+  private push(context: Context<unknown, unknown & Memory>): void {
+    this.pending.push(context);
+  }
+  static shouldStep(context: Context<unknown, unknown & Memory> | undefined, now: number): context is Context {
+    if (!context) return false;
+    if (context.updated === undefined) return true;
+    if (!context?.interval) return true;
+    return now - context.updated >= context.interval;
+  }
+  /**
+   * tick. Process all pending contexts, move them to done if needed
+   *
+   * @param {} now - Specify the timestamp this tick is running on.
+   * @returns {Promise<boolean>} - Returns if there are still any pending contexts to run.
+   */
+  tick = async (now = Date.now()): Promise<boolean> => {
+    const context = this.pop();
+    if (StateMachine.shouldStep(context, now)) {
+      let next;
+      switch (context.type) {
```

Comment:
we now switch on context.type to know how to cast parameters and memory of each context type before it goes into the handler